### DISCUSSION TITLE
WT32-ETH01 + mcp23017

### DISCUSSION THREAD

**Discussion by pin32** [arendst/Tasmota#19679] [Q&A] [answered]
good evening guys. I have a wt32-eth01 module with firmware 9.5, two questions
1) Is it upgradeable or is it custom firmware?
2) can the i2c bus be used with this fw to control the mcp23017?
thanks for your kind response.

**Comment by Jason2866** [comment]
GPIO32 and 33 can be used for i2c. So yes flashing an actual Tasmota32 version should work.
Ethernt driver is in every Tasmota32 variant build.

**Comment by sfromis** [comment]
If what you mean is that the board already has a regular build of Tasmota 9.5, sure it can be upgraded. Usability of I2C depends on what specific build, but while most builds include I2C, MCP23017 support is only with a custom build.

**Comment by Jason2866** [comment]
Correct MCP230XX is in no precompiled Tasmota firmware. https://github.com/arendst/Tasmota/blob/development/BUILDS.md

**Comment by pin32** [comment]
Thanks for the reply.
I tried with the latest precompiled version but the wt32 module is not present in the options for configuration on the device's webgui, while by flashing version 9.5 this possibility is present. I'm a little confused, how should I proceed to get the wt32 module with the related GPIO configurations? having support for mcp23017, there will probably be something that escapes me, perhaps with the updated fw (13.1) ethernet support is included directly and it is not necessary to choose the wt32 module in the device menu? in this case should I compile a custom version for the mcp23017 based on the updated fw?
Can you confirm which way to go.
Thanks for the reply

**Reply by sfromis** [reply]
What I'd do would be to first install the precompiled current version of Tasmota using https://tasmota.github.io/install/ and get the basic configuration of Wifi details right. Then apply the template to get Ethernet working. https://templates.blakadder.com/wireless_tag_WT32-ETH01.html

After that, do a custom build to include the MCP23017 support, and upgrade to the resulting binary. https://tasmota.github.io/docs/Compile-your-build/

And of course, config I2C on GPIO32 and 33 and wire it up.

Note that if you are using a working Tasmota 9.5 already, you can also upgrade from that, instead of doing the wired flashing.

**Comment by pin32** [comment]
in the meantime I did a test:
I used the latest fw (13.1) to see if the hardware responded well, the device works.
then I compiled a custom version with ethernet+mcp230xx+temperature sensors support, I loaded it and it seems to go well... or almost since the LAN port doesn't work, or at least the LEDs don't light up and there is no trace of the connection on the console of the LAN cable. I tested the "ethaddress 1" command and I get a response and a device reboot:

![immagine](https://github.com/arendst/Tasmota/assets/107401551/f72a1c2f-97d1-4ab2-ba35-2a807861384a)

as regards the i2c bus on the module, gpio 32 and 33 are not reported, the only ones that seem usable are gpio14 and 15
![immagine](https://github.com/arendst/Tasmota/assets/107401551/440bb741-976e-4c7b-8ae9-2a670b0e54e6)

this is in custom template that I configured

I didn't understand the use of the following:
{"NAME":"WT32_ETH01","GPIO":[1,1,1,1,1,1,0,0,1,0,1,1,3840,576,5600,0,0,0, 0.5568,0.0.0.0.0.0.0.0.1.1.0.1.1.0.0.1],"FLAG":0,"BASE":1}

**Reply by sfromis** [reply, marked as answer]
You **need** the gpios from the template for Ethernet support, and use the guide from the how-to link to the left on the page. And do not type it manually, but use copy/paste. What you posted here has syntax errors looking like manual typing.

Your screenshot here is not what you may have done as a custom template, but what you're doing in "Configure module" after having some custom template applied.

Also note that the template has specific instructions on pins for I2C, matching what Jason told you. According to that, I2C can **only** work on gpio 32/33, when using Ethernet.

**Comment by pin32** [comment]
ok. I understand, the string configures the device... it wasn't clear to me.
I did everything you advised me, the only thing that remains unclear to me is which terminals the I2C refers to, those terminals are not on the board. the reserved pins will probably have been used (such as rx and tx used for programming) but I have no idea where to find this information.

![immagine](https://github.com/arendst/Tasmota/assets/107401551/5fd878fc-daf7-4727-91e6-105d52c5faf3)
maybe this can help
https://files.seeedstudio.com/products/102991455/WT32-ETH01_datasheet_V1.1-%20en.pdf
the "CFG" pin (pin 2) connects to gpio 32 -scl-
the "485_EN" pin (pin 3) connects to gpio 33 -sda-
it could be correct... can you confirm for me?
Thank you

**Comment by pin32** [comment]
![immagine](https://github.com/arendst/Tasmota/assets/107401551/dd8c0edb-9415-4258-a59e-907eee217472)

everything seems to be working fine....
I haven't tried the i2c and pwm, I'll update as soon as I have news.

Is it possible to also configure the IP?
![immagine](https://github.com/arendst/Tasmota/assets/107401551/f86b68ea-f641-4a9b-ae92-2f70919a56ed)

**Comment by pin32** [comment]
![immagine](https://github.com/arendst/Tasmota/assets/107401551/bca4a714-d5db-47d3-bd04-d727432b1e81)

mcp23017 also works perfectly.

all that remains is to understand if it is possible to configure a static IP also for the Ethernet.

thank you 1000 for the support!!!

**Reply by sfromis** [reply]
See the docs. https://tasmota.github.io/docs/Commands/#ethernet

**Comment by pin32** [comment]
a small problem... but what if two mcp23017 are needed?

**Reply by arendst** [reply]
Give it a different address and daisy chain by i2c

**Comment by pin32** [comment]
from tests carried out, only the address 0x20 is recognised, I tried with 0x21 and 0x27 but it is not seen as mcp 23017 and in the case 0x27 as a humidity sensor (0x27 is quite normal), i2cscan sees the new address but then it doesn't work and the sensor 29 doesn't works.
I haven't tried connecting them both at the same time (0x20 and 0x21) as I don't have the second chip yet.
in the compilation, however, I see that the device address is indicated as "#define USE_MCP230xx_ADDR 0x20", I tried to indicate differently:
#define USE_MCP230xx_ADDR 0x20
#define USE_MCP230xx_ADDR 0x21
in the compilation but in the specific case only 0x21 is seen, I also tested
#define USE_MCP230xx_ADDR 0x2x
#define USE_MCP230xx_ADDR 0xxx
#define USE_MCP230xx
but the compilation fails.
I also tested the dual bus configuration (gpio14 and 15) but this doesn't seem to work either.
am I missing something?
Thanks for the reply.

**Reply by arendst** [reply]
yes you're missing something. Dive into the tasmota mcp23xxx docs and look for mode 2.

BTW this only works when both chips are connected AND a correct config is provided as documented.

**Comment by pin32** [comment]
it's quite clear that I missed something along the way.... but having already become familiar with the single mcp I went from memory...
Mode 2 is certainly very interesting and offers extended and very professional granularity (it's no surprise now and I know everyone that the Tasmota project is a step ahead of everything else).
the problem is that I'm not very behind it... I'm a 58 year old analogue electronics guy in the process of emancipation...
so if I understood correctly I compile with these definitions:

#define USE_MCP23XXX_DRV
#define USE_MCP230xx_ADDR 0x20

then I should write a configuration file or a rule to configure the mcp chip, but I don't quite understand what to insert in this file or rule to obtain 16 pullup inputs on one chip and 16 outputs on the other... maybe you can give me a little help ?

thank you 1000

**Comment by pin32** [comment]
if I understood correctly (definitely not)

this could be a 16 input configuration
{"NAME":"MCP23017 expander","BASE":0,"GPIO":[160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175]}

and this is a configuration for 16 outputs
{"NAME":"MCP23017 expander","BASE":0,"GPIO":[224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239]}

but I can't understand how they address the two MCPs, how do I decide which one and which the other? deep night in my mind....

**Reply by arendst** [reply]
You're on the correct way. Just add the gpios from the second config to the first config et presto. Think of it as configuring the gpios from lowest to highest i2c address.

To see if your config is correct enable more logging and do a restart. If configured correctly it should report the new buttons/switches.

**Comment by pin32** [comment]
it's useless... you're truly a GENIUS!!!
![immagine](https://github.com/arendst/Tasmota/assets/107401551/ae18d3e3-6fdb-402f-a0a5-b067f690a1d8)

seeing as today is a day of cultural growth (for me)... I just need to understand the use of "BASE"...
A=first section of the first mcp
B=second section of the first mcp
C=first section of the second mcp
D=second section of the second mcp
and here:
BASE: A=S1-8, B=S9-16 C=R1-8, D=R9-16
If I understand correctly you get the same result with:

{"NAME":"MCP23017 expander","BASE":[A=S1-8, B=S9-16 C=R1-8, D=R9-16],"GPIO":0}

it's right?

**Reply by arendst** [reply]
Nope. Try it and it fails. Base is a flag. telling the code to start numbering the mcp inputs from 0 or adding from the already present esp gpios if defined.

**Reply by salvo76** [reply]
Hi Pin32, if a strength, could I ask you for some help to achieve the same thing as you? I'll update you on my progress:
1) Compiled firmware with custom with tasmocompiler and added the string #define USE_MCP23XXX_DRV.
2) Flash the WT32-ETH01 and I got both lan and wifi.
3) Assigned on pins 32-33 SCL-SDA, from Tasmota console I run I2CScan and get: 17:29:25.845 MQT: stat/tasmota_880938/RESULT = {"I2CScan":"Device(s) found on bus1 at 0x20 0x21" }, so I think everything is ok, the two MCP23017 are detected by the I2C bus. But then I can no longer go forward to achieve what you did, that is, 16 inputs on the first MCP and 16 outputs on the second.
Could you please help me?

**Reply by sfromis** [reply]
You need to select mode 2 in your custom build to enable the expander support at all, and to set up the special template for configuring the pins.
https://tasmota.github.io/docs/MCP230xx/#mode-2

**Reply by salvo76** [reply]
Yes, I selected the mode by adding #define USE_MCP23XXX_DRV, I don't understand where to put the special model for the pin configuration

**Reply by sfromis** [reply]
That is explained in the link I posted.

> if a filesystem is present it looks for file mcp23x.dat

**Reply by salvo76** [reply]
In the meantime, thank you for your availability, I have read and reread this page, but I cannot understand why the MCP23X.dat file is not found by the console. How do I create it? I have enabled mode 2.

**Reply by barbudor** [reply]
If you haven't created it, there's no way it will find it
Main Menu => Consoles => Manage file system => Create and edit a new files
Dont forget to change the name, paste the configuration JSON you need, save and restart

**Reply by salvo76** [reply]
Thanks, now everything is clear and I have also solved it, now I will try to perfect it for my needs, thanks everyone
![Cattura](https://github.com/arendst/Tasmota/assets/61916852/70d970ac-e8a2-4d4d-9280-a02f5e0d2786)

**Comment by pin32** [comment]
in my specific case that I have three switches and three relays already configured the GPIOs of the two MCPs would be placed after these.

{"NAME":"MCP23017 expander","BASE":[A=S4-11, B=S12-19 C=R4-11, D=R12-19]"GPIO":[160,161,162,163,164,165,166,167,168,169,170,171,172,173,174, 175,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239]}

Is the syntax correct?

**Reply by sfromis** [reply]
Minor detail, you're missing a comma before `"GPIO"`. And the value of `BASE` is **only** a flag. What you have there looks more like what can be included in the `NAME` parameter, as illustrated by the examples in https://tasmota.github.io/docs/MCP230xx/#mode-2

**Reply by barbudor** [reply]
No

You didn't read the doc carefully and Theo confirmed that BASE is a number
```
an optional "BASE" data pair selecting if either relative (0 = default) or absolute (1) button and/or switch numbering is used
```

With "BASE": 0, the MCP relays and buttons/swictches are numbered from 1 and the native ones comes after
With "BASE": 1, the native relays and buttons/switches are numbered from 1 and the MCP ones comes after

**Comment by pin32** [comment]
{"NAME":"MCP23017 expander","BASE":[A=S4-11, B=S12-19 C=R4-11, D=R12-19],"GPIO":[160,161,162,163,164,165,166,167,168,169,170,171,172,173,174, 175,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239]}

{"NAME":"MCP23017 expander","BASE":1,"GPIO":[160,161,162,163,164,165,166,167,168,169,170,171,172,173,174, 175,224,225,226,227,228,229,230, 231,232,233,234,235,236,237,238,239]}

If I understand correctly these are equivalent  (I corrected the syntax)

**Reply by barbudor** [reply]
> "BASE":[A=S4-11, B=S12-19 C=R4-11, D=R12-19]"

This is not equivalent to anything, this is just totally wrong and pure invention.

**Comment by pin32** [comment]
It's actually clear in the guide...
![immagine](https://github.com/arendst/Tasmota/assets/107401551/1bbc61ee-8b74-4f34-8b92-da9233e3678e)

you can omit "BASE" and specify in "NAME" with the above syntax...

well now it's very clear, thanks for your patience!

**Reply by barbudor** [reply]
You don't "specify" anything in NAME, you just give a name that can be used as a reminder of the configuration.
But whatever you put in NAME as NO effect on the configuration
It's just a NAME

**Comment by pin32** [comment]
ok. at most it will be part of the name... a descriptive name.
Thank you

**Comment by pin32** [comment]
![immagine](https://github.com/arendst/Tasmota/assets/107401551/0b0f06d9-87e2-45dd-871b-9b9f09350458)

but here it is a bit confusing, there is the possibility of inserting data into "BASE"...

**Reply by barbudor** [reply]
0 or 1
These are the only documented possible values

**Reply by sfromis** [reply]
No data in "NAME". It is just a string, which you may use for readability, as the list of numbers in "GPIO" is not very easy to "just read". Nothing to suggest that "BASE" can contain more than 0 for relative, and 1 for absolute numbering.

**Comment by pin32** [comment]
everything perfect works great.
I'm having a couple of problems (maybe a little off topic)
- the "webbutton" command on relay 17, 18, 19 does not work, from 1 to 16 it does
- changing the configuration from "rele" to "rele_i" has no effect (at least on direct gpios, those connected directly to the chip), I don't know if it's possible on remote gpios (on mcp)
- 
![immagine](https://github.com/arendst/Tasmota/assets/107401551/5be3e3d9-8633-445b-8e3d-de49f25172c4)

![immagine](https://github.com/arendst/Tasmota/assets/107401551/4eceb409-25c4-4a73-a307-34b520bc1ad6)

**Reply by barbudor** [reply]
The WebButton command has very recently been improved to support values over 16.read carefully the doc as values are not saved.

**Comment by pin32** [comment]
the ds18x20 sensor doesn't seem to work either.
I tested several GPIOs and several (new) sensors.
maybe there are some bugs?

**Reply by Jason2866** [reply]
The DS18B20 does work with the ESP32. Which GPIO tried? You have not many left you can use. LAN8720 does need a lot of GPIOs. This GPIOs are used from LAN chip

  * Dedicated fixed Phy pins
 * GPIO17 - EMAC_CLK_OUT_180
 * GPIO19 - EMAC_TXD0(RMII)
 * GPIO21 - EMAC_TX_EN(RMII)
 * GPIO22 - EMAC_TXD1(RMII)
 * GPIO25 - EMAC_RXD0(RMII)
 * GPIO26 - EMAC_RXD1(RMII)
 * GPIO27 - EMAC_RX_CRS_DV
*  WT32 specific used
 * GPIO16 = Force Hi
 * GPIO18 = ETH MDIO
 * GPIO23 = ETH MDC

**Reply by sfromis** [reply]
Did you make sure to have a pullup resistor wired for the DS18B20?

**Comment by pin32** [comment]
![immagine](https://github.com/arendst/Tasmota/assets/107401551/efc69bfe-a0d9-4a33-9f06-0c9c2b33784e)
This is the configuration I use. I also tested other gpios but it is not detected.

**Comment by pin32** [comment]
![immagine](https://github.com/arendst/Tasmota/assets/107401551/ddd13bfc-0753-4ce9-8ea7-794be7b7beb2)
everything works great... I forgot about the pullup resistor....

**Reply by salvo76** [reply]
Hi, you're great, I've been trying to do something similar to yours for days, I configured the WT32-ETH01 with Tasmota, enabled I2C on the correct pins, from i2cscan I find the two addresses, but I don't know how to continue, could you help me?

**Reply by sfromis** [reply]
For I2C, you need to make sure to be using a build including support for your specific I2C devices, and sometimes you may have to disable drivers for other device types using same addresses. Standard Tasmtoa builds do not contain **all** sensors etc.

**Comment by pin32** [comment]
I would like to use the interrupt, in the guide I read:
![immagine](https://github.com/arendst/Tasmota/assets/107401551/29b42be3-57a2-44b2-8c24-0d933622c768)

I connected the two gpio pins 35 and 36 (which can be used as inputs) to the "IA and IB" outputs of the mcp addressed ox20 as 16 inputs and configured the micro as follows:
![immagine](https://github.com/arendst/Tasmota/assets/107401551/d8b5c912-fb4e-40ac-86c7-8c4d2fa44ce3)

I have verified that pins IA and IB generate the negated pulse correctly when the state of the inputs changes (with an oscilloscope), but I do not obtain any changes in the generation of the relevant interrupt data by monitoring from the console:
![immagine](https://github.com/arendst/Tasmota/assets/107401551/9bfd1086-4422-4ff6-9fe0-021889bd3412)

with the configuration without using the interrupt pins:
![immagine](https://github.com/arendst/Tasmota/assets/107401551/f177bfb7-849a-4391-83b8-820d5db2412d)
![immagine](https://github.com/arendst/Tasmota/assets/107401551/72272eb3-d84d-42cc-8574-b0f3fdd33c33)

I'm definitely missing some information... give me a little help. Thank you

**Comment by pin32** [comment]
I discovered that if you configure the inputs without pullup, the interrupt doesn't work... I didn't understand if this is normal behavior. during the tests by activating the interrupt pins I obtained a multitude of signals and checking the input pins they did not have the internal pullups inserted despite the correct configuration, so I decided to put them external and clearly I switched to switchmode without pullups... but in doing so I no longer got the int signals on the outputs of the mcp chip...
now with the configuration with internal pullups the output signals from the chip are present, but I have both internal and external pulls...
Can you please confirm if the configuration of the interrupts are linked to the configuration of the pullups?

**Reply by Jason2866** [reply]
There is no problem to have the internal weak pull ups additionally enabled.

**Comment by pin32** [comment]
Yes, of course. but the problem is that they don't activate and the interrupt doesn't seem to work either...

**Comment by pin32** [comment]
Using mode 2 with mcp230xx is a bit mysterious to me...

**Reply by barbudor** [reply]
Actually what are you expecting by using Interrupts ?
What it does is processing the input change faster (as soon as possible instead of every 100ms) and then notify the Button press or the Switch change.
So if you see your Button/Switches messages, then you get.

**Reply by barbudor** [reply]
After double checking I confirm that
- driver V1 (`#define USE_MCP230xx`) was sending message `MCP230XX_INT:`
- driver V2 (`#define USE_MCP230xx_DRV`) is no more sending such messages but handle the interrupts and generate Button/Switch messages

**Comment by pin32** [comment]
it was clear that with mode 2 the functions were simplified, but I didn't understand if the interrupt produces a telemetry update or if it publishes a particular/customized message

**Reply by sfromis** [reply]
The good thing about using mode 2 is that it just acts as standard Switch/Button inputs, and Relay outputs, instead of having to deal with very different messages. What specifically happens depends on how you configure the inputs, where getting messages is one option, you can couple the inputs with default relay actions, or you can set up to use rules to handle the inputs. Just like with any other Button/Switch input. https://tasmota.github.io/docs/Buttons-and-Switches/

**Comment by Jeziel-Guiev** [comment]
https://northvelia.com/introduccion-al-chip-ethernet-lan8720-con-el-wt32-eth01/

**Comment by MSc1988** [comment]
Hello everyone,
i use 3 MCP23017 on 0x20 0x21 0x22 in mode 2 (only driver 77 activ)
i set up this mcp23.dat

{"NAME":"MCP23017","GPIO":[256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279,0,0,0,0,0,0,0,0,32,33,34,35,36,37,38,9,160,161,0,0,0,0,0,0]}

so fare (after about 10 hours with different test) i did NOT manage to get the input right
for me it seems like, If I use a button or switch there is a rule behind that also turns a output relay (only mcp chip is connected via bus, no relay behind).
as I checked for rules, there are no rules.
new flashed tasmota 14.3.0 set up for test on esp32s3 and wt32-eth01, both esp32 have the same problems.
I did not find a way how to turn of behaviour of switching outputs when giving input.

when I use a button for example i get this output on loglevel 4:

22:38:18.531 BTN: Button5 multi-press 1
22:38:19.284 SRC: Button
22:38:19.297 RSL: RESULT = {"POWER5":"ON"}
22:38:19.297 RSL: POWER5 = ON
22:38:19.915 CFG: Saved, Count 57, Bytes 4096

what I did:
connect A4 to GND on 0x22
get a toggle A4 of =x20

may someone has a idea? 
thanks in advance!

**Reply by barbudor** [reply]
Please doc section 'Buttons and Switches'
https://tasmota.github.io/docs/Buttons-and-Switches/
By default buttons and switches are natively coupled to relays without the need of a rule. That is a base Tasmota behavior.

The chapter above list various methods to decouple buttons and switches from relays.

**Reply by MSc1988** [reply]
thank you very much for that advice, this solved my problem !!!
(need to detach button and switch seperatly)
I didn´t expected tasmota is that much advanced that it has nativ attachmend from input to output
